core/peerset/src/peersstate.rs

```diff
@@ -187,21 +192,160 @@ impl PeersState {
 				}
 				Some(to_try)
 			})
-			.map(|(peer_id, state)| (peer_id.clone(), state));
+			.map(|(peer_id, _)| peer_id.clone());
 
-		if let Some((peer_id, state)) = outcome {
+		if let Some(peer_id) = outcome {
 			Some(NotConnectedPeer {
-				state,
+				state: self,
 				peer_id: Cow::Owned(peer_id),
-				num_in: &mut self.num_in,
-				num_out: &mut self.num_out,
-				max_in: self.max_in,
-				max_out: self.max_out,
 			})
 		} else {
 			None
 		}
 	}
+
+	fn disconnect(&mut self, peer_id: &PeerId) {
+		let is_priority = self.is_priority(peer_id);
+		if let Some(mut node) = self.nodes.get_mut(peer_id) {
+			if !is_priority {
+				match node.connection_state {
+					ConnectionState::In => self.num_in -= 1,
+					ConnectionState::Out => self.num_out -= 1,
+					ConnectionState::NotConnected =>
+						debug_assert!(false, "State inconsistency: disconnecting a disconnected node")
+				}
+			}
+			node.connection_state = ConnectionState::NotConnected;
+		} else {
+			warn!(target: "peerset", "Attempting to disconnect unknown peer {}", peer_id);
+		}
+	}
+
+	/// Sets the peer as connected with an outgoing connection.
+	pub fn try_outgoing(&mut self, peer_id: &PeerId) -> bool {
```

Comment:
I'm not sure why this method's body was moved here and didn't stay in the regular `try_outgoing` method, but that method should be private.

Reply:
It was moved, because the the `NotConnectedPeer` does not hold all the data required for this function anymore. It is a function of shared state now.